### PR TITLE
[14.0][ADD] stock_barcodes_serial_source_relocation: new module

### PR DESCRIPTION
**Pull request by FrankC013**

### CODE REVIEW

**Review by eantones — APPROVED** [review]
Good job!!

File: stock_barcodes_serial_source_relocation/models/stock_barcodes_read.py [hunk 1]
```diff
@@ -55,9 +70,44 @@ def action_done(self):
...
                 return False
             if existent_quant:
-                move_quant = self.env["stock.move"].create(
-                    self._prepare_relocation_move_values(existent_quant)
+                picking_type = self.env["stock.picking.type"].search(
+                    [
+                        (
+                            "warehouse_id",
+                            "=",
+                            self.picking_id.picking_type_id.warehouse_id.id,
```
Comment (eantones) [inline stock_barcodes_serial_source_relocation/models/stock_barcodes_read.py]: get_warehouse and check if the location has a warehouse

File: stock_barcodes_serial_source_relocation/models/stock_picking_type.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class PickingType(models.Model):
+    _inherit = "stock.picking.type"
+
+    is_regularization = fields.Boolean(copy=False)
+
+    @api.constrains("is_regularization")
```
Comment (eantones) [inline stock_barcodes_serial_source_relocation/models/stock_picking_type.py]: add warehouse_id and the rest

File: stock_barcodes_serial_source_relocation/models/stock_picking_type.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        for rec in self:
+            if rec.is_regularization:
+                if rec.code != "internal":
+                    raise ValidationError(
+                        _("Only internal picking types can be regularization."),
+                    )
+                if (
+                    rec.env["stock.picking.type"].search_count(
+                        [
+                            ("is_regularization", "=", True),
```
Comment (eantones) [inline stock_barcodes_serial_source_relocation/models/stock_picking_type.py:24]: add code

File: stock_barcodes_serial_source_relocation/models/stock_barcodes_read.py [hunk 4]
```diff
@@ -55,9 +70,44 @@ def action_done(self):
...
+                        ("code", "=", "internal"),
+                        ("is_regularization", "=", True),
+                    ]
+                )
+                if not picking_type:
+                    picking_type = self.env["stock.picking.type"].search(
+                        [
+                            ("warehouse_id", "=", False),
+                            ("code", "=", "internal"),
+                            ("is_regularization", "=", True),
```
Comment (eantones) [inline stock_barcodes_serial_source_relocation/models/stock_barcodes_read.py]: add company

**Review by eantones — CHANGES_REQUESTED** [review]

**Comment by FrankC013** [comment]
@eantones This morning's changes

File: stock_barcodes_serial_source_relocation/models/stock_picking.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+                            .search(
+                                [
+                                    ("warehouse_id", "in", (warehouse.id, False)),
+                                    ("code", "=", "internal"),
+                                    ("is_regularization", "=", True),
+                                ],
+                            )
+                            .sorted(
+                                lambda x: x.warehouse_id.id, reverse=True
+                            )  # provar sense id # test
```
Comment (eantones) [inline stock_barcodes_serial_source_relocation/models/stock_picking.py]: Is this resolved??

File: stock_barcodes_serial_source_relocation/models/stock_picking.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+                        )
+                        warehouse_ids = picking_type.mapped(
+                            "warehouse_id.id"
+                        )  # veure que retorna, fer diferents proves
```
Comment (eantones) [inline stock_barcodes_serial_source_relocation/models/stock_picking.py]: Is this resolved? Please don't leave this comments like these... Unattended and not in English

File: stock_barcodes_serial_source_relocation/models/stock_picking.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+                        q.quantity,
+                        0,
+                        precision_rounding=move_line.product_id.uom_id.rounding,
+                    )
+                    > 0
+                )
+                if len(quants) > 1:
+                    raise ValidationError(
+                        _(
+                            "The S/N %s has quantities in multiple locations"  # modify
```
Comment (eantones) [inline stock_barcodes_serial_source_relocation/models/stock_picking.py]: Finish this "modify"

File: stock_barcodes_serial_source_relocation/models/stock_picking.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+                            )
+                        new_picking.button_validate()
+                    else:
+                        inventory = self.env["stock.inventory"].create(
+                            self._prepare_relocation_inventory_values(move_line)
+                        )
+                        inventory._action_start()
+                        inventory.with_context(
+                            relocation_origin=self.name
+                        ).action_validate()
```
Comment (eantones) [inline stock_barcodes_serial_source_relocation/models/stock_picking.py]:
```suggestion
                if quants:
                    qty_available = quants.filtered(
                        lambda x: x.location_id == move_line.location_id
                    ).quantity
                    if (
                        float_compare(
                            move_line.qty_done,
                            qty_available,
                            precision_rounding=move_line.product_id.uom_id.rounding,
                        )
                        > 0
                    ):
                        warehouse = move_line.location_id.get_warehouse()
                        if not warehouse:
                            raise ValidationError(
                                _(
                                    "No warehouse found for the location %s"
                                    % move_line.location_id.name
                                )
                            )
                        picking_type = (
                            self.env["stock.picking.type"]
                                .search(
                                [
                                    ("warehouse_id", "in", (warehouse.id, False)),
                                    ("code", "=", "internal"),
                                    ("is_regularization", "=", True),
                                ],
                            )
                                .sorted(
                                lambda x: x.warehouse_id.id, reverse=True
                            )  # provar sense id # test
                        )
                        warehouse_ids = picking_type.mapped(
                            "warehouse_id.id"
                        )  # veure que retorna, fer diferents proves
                        if len(warehouse_ids) != len(set(warehouse_ids)):
                            raise ValidationError(
                                _(
                                    "More than one regularization picking "
                                    "type for the same warehouse %s"
                                    % move_line.location_id.name
                                )
                            )
                        if not picking_type:
                            raise ValidationError(
                                _(
                                    "No regularization picking type for location %s"
                                    % move_line.location_id.name
                                )
                            )
                        new_picking = self.env["stock.picking"].create(
                            self._prepare_relocation_picking_values(
                                move_line, picking_type[0], quants
                            )
                        )
                        new_picking.action_confirm()
                        for move in new_picking.move_lines:
                            move.move_line_ids.write(
                                self._prepare_relocation_move_line_values(
                                    move_line, new_picking, quants
                                )
                            )
                        new_picking.button_validate()
                else:
                    inventory = self.env["stock.inventory"].create(
                        self._prepare_relocation_inventory_values(move_line)
                    )
                    inventory._action_start()
                    inventory.with_context(
                        relocation_origin=self.name
                        ).action_validate()

```
It's clearer and cleaner this way

**Review by eantones — CHANGES_REQUESTED** [review]

**Review by eantones — CHANGES_REQUESTED** [review]
On Monday, please squash all the commits into a single one

**Comment by FrankC013** [comment]
> On Monday, please squash all the commits into a single one

@eantones Done!

**Review by eantones — APPROVED** [review]

**Review by eantones — APPROVED** [review]

**Review by eantones — APPROVED** [review]
Good job!! 👌

**Pull request merged**